klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,165 @@
+        logger.info("%s was deleted; removing from klaus", event.src_path)
+        self.app.update_repos_list()
+
+    def on_moved(self, event):
+        """
+        Moved is not the same as deleted, unless the moved dir is no longer a subset of
+        any of the watched trees. In that case we need to consider it a deletion,
+        and remove it from the list of watched repos; otherwise we need to update its path.
+        This is a potentially expensive operation, because the only reliable way
+        to detect whether it is a subset (including symlinks etc) is to walk the path.
+        In practice, this shouldn't affect performance in the general case,
+        as moving a repository is a rare event.
+        """
+
+        matching = filter(lambda existing_repo : existing_repo.path == event.src_path, self.app.repos)
+        if not matching:
+            logger.debug("ignored non-repo move, %s -> %s", event.src_path, event.dest_path)
+            return
+
+        #first unregister old repositories
+        map(lambda x: self.app.repos.remove(x), matching)
+        # then see if the new destination is part of the klaus watchlist
+        for directory in self.app.basedirs:
+            if is_subdirectory(directory, event.dest_path):
+                logger.info("'%s' -> '%s'; updating references", event.src_path, event.dest_path)
+                self._add_dir(event.dest_path)
+
+        self.app.update_repos_list()
+        logger.info(
+            "%s was moved out of KLAUS_REPOS to %s, and is no longer a watched repository",
+            event.src_path,
+            event.dest_path
+        )
+
+
+class KlausAutoDiscover(Klaus):
+    """
+    Instead of the normal Klaus behaviour of receiving a list of git repositories
+    as `os.environ['KLAUS_REPOS']`, this app searches `os.environ['KLAUS_REPOS']`
+    for valid git repositories and updates the application automatically when any
+    changes are made.
+    This simplifies administration significantly, but may cause problems with
+    large repositories.
+    """
+    def __init__(self, basedirs, site_name, use_smarthttp, **kwargs):
+        self.basedirs = basedirs
+        repo_paths = []
+        # first find all the existing repos in the given base directory
+        for directory in basedirs:
+            for fspath in os.listdir(directory):
+                try:
+                    path = os.path.join(directory, fspath)
+                    FancyRepo(path)  # The result is discarded. This is just a test.
+                    repo_paths.append(path)
+                except dulwich.errors.NotGitRepository:
+                    continue
+        self.observer = Observer()
+        fs_event_handler = KlausEventHandler(self)
+        # Then setup up a handler to notify us when the content of any of those directories change.
+        map(lambda directory: self.observer.schedule(fs_event_handler, directory, recursive=True), basedirs)
```

Comment:
Please don't use `map` for functions that have side effects.

Also, if I understand this correctly, the watching goes recursively without a maximum depth level. This is going to be a problem as repositories may have very deeply nested directory hierarchies. Could the recursion level be limited somehow? Maybe we don't even need recursion. Maybe we're fine with only looking at the first level? I understand people may have complex repository hierarchies. But klaus doesn't need to support every extreme border case.